### PR TITLE
[fix] Add a cluster environment teardown to clean up environment state

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Discussion in #6853 
Fixes part of #6303

https://github.com/PyTorchLightning/pytorch-lightning/blob/f852a4f5925211d6bb9dac231b9185e3e6814b13/pytorch_lightning/plugins/training_type/ddp.py#L283-L285

This is a bug for users who have subsequent calls to 
```
trainer.fit(...)
trainer.test(...)
```
and who launched with torchelastic. The world size environment variable is mistakenly removed, causing the trainer.test call to fail. As a result, we introduce a `teardown` hook on the cluster environment to resolve this issue. the teardown for torchelastic and slurm are no-ops, but for the lightning environment, we can remove the world size after the sub-process calls finish.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 1]
```diff
@@ -280,7 +280,10 @@ def pre_dispatch(self):
 
         self.barrier()
 
-    def post_dispatch(self):
+    def post_dispatch(self) -> None:
+        # If the plugin launched subprocesses, clean up the populated environment variable(s)
+        if self.cluster_environment.creates_children:
+            return
         if "WORLD_SIZE" in os.environ:
             del os.environ["WORLD_SIZE"]
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp.py:288]: how about we ask the cluster env to clean up?

self.cluster_environment.clean_up / teardown

Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:288]: yes totally agreed

**Review by awaelchli — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ## [1.3.0] - 2021-MM-DD
 
 ### Added
+- Added a `teardown` hook to `ClusterEnvironment` ([#6942](https://github.com/PyTorchLightning/pytorch-lightning/pull/6942))
```
Comment (carmocca) [inline CHANGELOG.md:11]:
```suggestion


- Added a `teardown` hook to `ClusterEnvironment` ([#6942](https://github.com/PyTorchLightning/pytorch-lightning/pull/6942))
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -189,6 +191,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 
 ### Fixed
+- Fixed incorrect removal of `WORLD_SIZE` environment variable in DDP training when launching with torch distributed/torchelastic ([#6942](https://github.com/PyTorchLightning/pytorch-lightning/pull/6942))
```
Comment (carmocca) [inline CHANGELOG.md:194]:
```suggestion


- Fixed incorrect removal of `WORLD_SIZE` environment variable in DDP training when launching with torch distributed/torchelastic ([#6942](https://github.com/PyTorchLightning/pytorch-lightning/pull/6942))
```

**Review by carmocca — APPROVED** [review]
